warehouse/schema/schema.go

```diff
@@ -78,56 +77,89 @@ type schema struct {
 	fetchSchemaRepo                  fetchSchemaRepo
 	now                              func() time.Time
 	cachedSchema                     model.Schema
-	cacheExpiry                      time.Time
 	cachedSchemaMu                   sync.RWMutex
 }
 
 func New(
-	db *sqlquerywrapper.DB,
+	ctx context.Context,
 	warehouse model.Warehouse,
 	conf *config.Config,
-	logger logger.Logger,
+	slogger logger.Logger,
 	statsFactory stats.Stats,
 	fetchSchemaRepo fetchSchemaRepo,
-) Handler {
+	schemaRepo schemaRepo,
+	stagingFileRepo stagingFileRepo,
+) (Handler, error) {
 	ttlInMinutes := conf.GetDurationVar(720, time.Minute, "Warehouse.schemaTTLInMinutes")
-	schema := &schema{
+	sh := &schema{
 		warehouse:                        warehouse,
-		log:                              logger.Child("schema"),
+		log:                              slogger.Child("schema"),
 		ttlInMinutes:                     ttlInMinutes,
-		schemaRepo:                       repo.NewWHSchemas(db),
+		schemaRepo:                       schemaRepo,
 		stagingFilesSchemaPaginationSize: conf.GetInt("Warehouse.stagingFilesSchemaPaginationSize", 100),
-		stagingFileRepo:                  repo.NewStagingFiles(db),
+		stagingFileRepo:                  stagingFileRepo,
 		fetchSchemaRepo:                  fetchSchemaRepo,
 		enableIDResolution:               conf.GetBool("Warehouse.enableIDResolution", false),
 		now:                              timeutil.Now,
 	}
-	schema.stats.schemaSize = statsFactory.NewTaggedStat("warehouse_schema_size", stats.HistogramType, stats.Tags{
+	sh.stats.schemaSize = statsFactory.NewTaggedStat("warehouse_schema_size", stats.HistogramType, stats.Tags{
 		"module":        "warehouse",
 		"workspaceId":   warehouse.WorkspaceID,
 		"destType":      warehouse.Destination.DestinationDefinition.Name,
 		"sourceId":      warehouse.Source.ID,
 		"destinationId": warehouse.Destination.ID,
 	})
-	return schema
+	// cachedSchema can be computed in the constructor
+	// we need not worry about it getting expired in the middle of the job
+	// since we need the schema to be the same for the entireduration of the job
+	whSchema, err := sh.schemaRepo.GetForNamespace(
+		ctx,
+		sh.warehouse.Source.ID,
+		sh.warehouse.Destination.ID,
+		sh.warehouse.Namespace,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("getting schema for namespace: %w", err)
+	}
+	if whSchema.Schema == nil {
+		sh.cachedSchema = model.Schema{}
+		return sh, nil
+	}
```

Comment:
`whSchema.Schema == nil` When can sh.schemaRepo.GetForNamespace return nil? And what would nil value mean?

Reply:
It will be `nil` in case if no entry exists in RudderStack's postgres for that namespace
https://github.com/rudderlabs/rudder-server/blob/e711579eac45033cd8138cbd24172b15aae93bd0/warehouse/internal/repo/schema.go#L114

Reply:
In this case, why not call `fetchSchemaFromWarehouse`? As in for expired, we are fetching. Trying to understand difference between expired and nil.

Reply:
Yes no entry would mean that this is the first sync. Post first sync, there will always be an entry. And by entry, I mean the complete row and not just entry in schema column.